app/src/main/java/ch/epfl/cs311/wanderwave/model/spotify/SpotifyController.kt

```diff
@@ -97,34 +99,95 @@ class SpotifyController(private val context: Context) {
-              remote.playerApi.subscribeToPlayerState().setEventCallback {
-                startPlaybackTimer(it.track.duration)
-              }
-              trySend(true)
-            }
-            .setErrorCallback {
-              Log.e("SpotifyController", "Failed to play track: ${track.title}")
-              trySend(false)
-            }
+  fun playTrack(track: Track, onSuccess: () -> Unit = {}, onFailure: (Throwable) -> Unit = {}) {
+    appRemote.value?.let {
+      it.playerApi
+          .play(track.id)
+          .setResultCallback { onSuccess() }
+          .setErrorCallback { error -> onFailure(error) }
+    }
+  }
+
+  fun playTrackList(
+      trackList: List<Track>,
+      track: Track? = null,
+      onSuccess: () -> Unit = {},
+      onFailure: (Throwable) -> Unit = {}
+  ) {
+    if (trackList.isEmpty()) {
+      onFailure(Throwable("Empty track list"))
+      return
+    }
+    val trackToPlay = track ?: trackList[0]
+    appRemote.value?.let {
+      it.playerApi
+          .play(trackToPlay.id)
+          .setResultCallback {
+            this.trackList = trackList
+            onSuccess()
+          }
+          .setErrorCallback { error -> onFailure(error) }
+    }
+  }
+
+  fun pauseTrack(onSuccess: () -> Unit = {}, onFailure: (Throwable) -> Unit = {}) {
+    appRemote.value?.let {
+      it.playerApi
+          .pause()
+          .setResultCallback { onSuccess() }
+          .setErrorCallback { error -> onFailure(error) }
+    }
+  }
+
+  fun resumeTrack(onSuccess: () -> Unit = {}, onFailure: (Throwable) -> Unit = {}) {
+    appRemote.value?.let {
+      it.playerApi
+          .resume()
+          .setResultCallback { onSuccess() }
+          .setErrorCallback { error -> onFailure(error) }
+    }
+  }
+
+  suspend fun skip(
+      direction: Int,
+      onSuccess: () -> Unit = {},
+      onFailure: (Throwable) -> Unit = {}
+  ) {
+    val currentTrack = playerState().firstOrNull()?.track
+    val currentIndex =
+        trackList?.indexOfFirst { track -> track.id == currentTrack?.uri } ?: -1
+    Log.d("SpotifyController", "Skipping $direction")
+    Log.d("SpotifyController", "Track list: $trackList")
+    Log.d("SpotifyController", "Current index: $currentIndex")
+    if (currentIndex != -1) {
+      val nextIndex = (currentIndex + direction) % trackList!!.size
+      val nextTrack = trackList!![nextIndex]
+      playTrack(nextTrack, onSuccess, onFailure)
+    }
+  }
+
+  @OptIn(ExperimentalCoroutinesApi::class)
+  fun playerState(): Flow<PlayerState?> {
+    return appRemote.flatMapLatest { appRemote ->
+      callbackFlow {
+        val callbackResult =
+            appRemote
+                ?.playerApi
+                ?.subscribeToPlayerState()
+                ?.setEventCallback {
+                  trySend(it)
+                  startPlaybackTimer(it.track.duration)
+                }
+                ?.setErrorCallback { Log.d("SpotifyController", "Error in player state flow") }
```

Comment:
(it's a detail, but i think Log.e is better for this kind of stuff)

Reply:
Fixed in the next commit